File: daily/Sliding Window/567-Permutation-in-string.go
package daily

import "reflect"

func checkInclusion(s1 string, s2 string) bool {
	if len(s1) > len(s2) {
		return false
	}
	targetWindow := make(map[rune]int)
	for _, char := range s1 {
		targetWindow[char]++
	}
	k := len(s1)
	currentWindow := make(map[rune]int)
	for i := 0; i < k; i++ {
		currentWindow[rune(s2[i])]++
	}

	for i := k; i < len(s2); i++ {
		if reflect.DeepEqual(targetWindow, currentWindow) {
			return true
		}
		currentWindow[rune(s2[i])]++
		currentWindow[rune(s2[i-k])]--
		if currentWindow[rune(s2[i-k])] == 0 {
			delete(currentWindow, rune(s2[i-k]))
		}
	}
	return reflect.DeepEqual(targetWindow, currentWindow)
}

File: daily/Sliding Window/567-Permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        target_window = Counter(s1)
        k = len(s1)
        current_window = defaultdict(int)
        for i in range(k):
            current_window[s2[i]] += 1
        
        for i in range(k, len(s2)):
            if target_window == current_window:
                return True
            current_window[s2[i]] += 1
            current_window[s2[i - k]] -= 1
            if current_window[s2[i - k]] == 0:
                del current_window[s2[i - k]]

        return target_window == current_window
